### ISSUE TITLE
[tests] move tests into src

### ISSUE THREAD

**Issue by wgtdkp** [tianyuxf/tins#2]
We should treat tests more serious and follow Java's practice of managing tests in `src/test`.

Personally, I will also recommend we follow [TDD](https://en.wikipedia.org/wiki/Test-driven_development) and write tests before implementation. Thoughts? @tianyuxf

**Comment by joey-wang97** [comment]
you're right, In fact, I wanted to do it lately. I will improve it as soon.

Actually, I had some problem  writing the `parser`. could you give me some advices?
See below:
```
prefixUnaryExpr = '-' factorExpr
                | '~' prefixUnaryExpr
                | '++' factorExpr
                | '--' factorExpr
                | '!' prefixUnaryExpr
                | '(' TYPE ')' prefixUnaryExpr # cast expr
                | factorExpr

factorExpr = factorExpr '.' ID
           | factorExpr '[' expr ']'
           | callFuncExpr
           | ID
           | NUM_VAL | STRING_VAL | CHAR_VAL | DOUBLE_VAL | FLOAT_VAL
           | '(' expr ')' # parenthesis expr
```
When I try to parser the `cast expr` in `prefixUnaryExpr`, If I meet a token sequence: `'(', ID, ') other token...'`, I can't immediately decide which grammer production should to use.

```
int i = 10
# its cast expr, the struct defined below is supported
object o = (node) i 
 #it's parenthesis expr here
object o = (n1) + 1
struct node{}
node n1
```
the `TYPE` in `cast expr` maybe not in symbols, just a ID, cause it will load lately.
And I think I can't read the `prefixUnaryExpr` in the tail of `cast expr`, because it maybe a part of other exprs, like `(n1)+1`. I meet this problem last night, could you give me some advices? 
Thanks a lot!

**Comment by wgtdkp** [comment]
> the TYPE in cast expr maybe not in symbols, just a ID, cause it will load lately.

In C language, if the type declaration is not present when the identifier is used, the identifier will be not be treated a `type` but simply an `identifier`.

Other languages may support "use before declaration", but that will require a different syntax or multiple passes of parsing (e.g. parse all declarations at the first pass and parse function bodies at the second pass).

For your project, I would suggest we keep it simple and require that the type must be declared before being used. Thoughts? @tianyuxf

**Comment by joey-wang97** [comment]
`Multiple passes` would be helped for me.

I also thought declared should before used at first, but after thinking and investigating, the developer maybe should not focus on the declared order of program. I think that is simple, at least for developer.

**Comment by wgtdkp** [comment]
okay, then I think you should go with `multiple passes`.
